File: sdk-python/devutils_sdk/core/job_resolver.py
"""Job resolver for async operations"""

import asyncio
from typing import Optional, Dict, Any
from dataclasses import dataclass


@dataclass
class JobResponse:
    """Response from job creation"""

    job_id: str
    status: str
    created_at: str
    expires_at: Optional[str] = None
    result: Optional[Dict[str, Any]] = None


@dataclass
class JobResolverConfig:
    """Configuration for job resolver"""

    poll_interval: int = 1000  # milliseconds
    max_wait_time: int = 300000  # milliseconds (5 minutes)
    timeout: int = 30  # seconds


class JobResolver:
    """Resolves async job results"""

    def __init__(self, config: Optional[JobResolverConfig] = None):
        self.config = config or JobResolverConfig()

    @staticmethod
    def has_job_id(response: Dict[str, Any]) -> bool:
        """Check if response contains job ID"""
        return "job_id" in response or "jobId" in response

    @staticmethod
    def get_job_id(response: Dict[str, Any]) -> str:
        """Extract job ID from response"""
        return response.get("job_id") or response.get("jobId")

    @staticmethod
    def is_job_complete(job_status: str) -> bool:
        """Check if job is complete"""
        return job_status in ("completed", "success", "done")

    @staticmethod
    def is_job_failed(job_status: str) -> bool:
        """Check if job failed"""
        return job_status in ("failed", "error", "cancelled")

    async def resolve(
        self,
        job_id: str,
        get_status_fn,
    ) -> Dict[str, Any]:
        """
        Resolve job result by polling

        Args:
            job_id: Job ID to resolve
            get_status_fn: Async function to get job status

        Returns:
            Job result

        Raises:
            TimeoutError: If job doesn't complete within max_wait_time
        """
        start_time = asyncio.get_event_loop().time()

        while True:
            elapsed = (asyncio.get_event_loop().time() - start_time) * 1000

            if elapsed > self.config.max_wait_time:
                raise TimeoutError(f"Job {job_id} did not complete within {self.config.max_wait_time}ms")

            try:
                status_response = await get_status_fn(job_id)

                if self.is_job_complete(status_response.get("status")):
                    return status_response

                if self.is_job_failed(status_response.get("status")):
                    raise RuntimeError(f"Job {job_id} failed: {status_response.get('error')}")

                await asyncio.sleep(self.config.poll_interval / 1000)

            except Exception as e:
                if isinstance(e, (TimeoutError, RuntimeError)):
                    raise
                await asyncio.sleep(self.config.poll_interval / 1000)

File: sdk/core/job-resolver.ts
/**
 * Job Resolver for DevUtils SDK
 * Polls async jobs (screenshot, PDF) until completion
 * Handles timeouts and exponential backoff
 */

import { retry, RetryConfig } from "./retry-engine";
import { DevUtilsError, normalizeError } from "./error-handler";

export interface JobResponse {
  jobId?: string;
  id?: string;
  status?: string;
  data?: any;
  success?: boolean;
  error?: any;
}

export interface JobResolverConfig {
  maxWaitMs?: number;
  initialPollDelayMs?: number;
  maxPollDelayMs?: number;
  pollBackoffMultiplier?: number;
}

export const DEFAULT_JOB_RESOLVER_CONFIG: Required<JobResolverConfig> = {
  maxWaitMs: 30000, // 30 seconds
  initialPollDelayMs: 500,
  maxPollDelayMs: 3000,
  pollBackoffMultiplier: 1.5,
};

/**
 * Check if response contains a job that needs polling
 */
function hasJobId(response: JobResponse): boolean {
  return !!(response.jobId || response.id);
}

/**
 * Get job ID from response
 */
function getJobId(response: JobResponse): string {
  return (response.jobId || response.id) as string;
}

/**
 * Check if job is complete
 */
function isJobComplete(jobStatus: string): boolean {
  return (
    jobStatus === "COMPLETED" || jobStatus === "DONE" || jobStatus === "SUCCESS"
  );
}

/**
 * Check if job failed
 */
function isJobFailed(jobStatus: string): boolean {
  return (
    jobStatus === "FAILED" || jobStatus === "ERROR" || jobStatus === "CANCELLED"
  );
}

/**
 * Sleep for a given number of milliseconds
 */
function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

/**
 * Resolve a job by polling until completion
 * If response doesn't have a jobId, returns immediately
 * @param response - Initial API response
 * @param pollFn - Function to poll job status
 * @param config - Job resolver configuration
 * @returns Final job data
 */
export async function resolveJob<T = any>(
  response: JobResponse,
  pollFn: (jobId: string) => Promise<JobResponse>,
  config: Partial<JobResolverConfig> = {},
): Promise<T> {
  const finalConfig = { ...DEFAULT_JOB_RESOLVER_CONFIG, ...config };

  // If no job ID, response is already complete
  if (!hasJobId(response)) {
    if (response.data) {
      return response.data as T;
    }
    return response as any as T;
  }

  const jobId = getJobId(response);
  const startTime = Date.now();
  let pollDelay = finalConfig.initialPollDelayMs;
  let pollAttempt = 0;

  while (true) {
    // Check timeout
    const elapsed = Date.now() - startTime;
    if (elapsed > finalConfig.maxWaitMs) {
      throw new DevUtilsError(
        "JOB_TIMEOUT",
        `Job ${jobId} did not complete within ${finalConfig.maxWaitMs}ms`,
        408,
      );
    }

    // Wait before polling
    if (pollAttempt > 0) {
      await sleep(pollDelay);
      // Increase delay for next poll (exponential backoff)
      pollDelay = Math.min(
        pollDelay * finalConfig.pollBackoffMultiplier,
        finalConfig.maxPollDelayMs,
      );
    }

    try {
      // Poll job status
      const jobResponse = await retry(() => pollFn(jobId), {
        maxRetries: 3,
        initialDelayMs: 100,
        maxDelayMs: 500,
      });

      const status = jobResponse.status || "UNKNOWN";

      // Job completed successfully
      if (isJobComplete(status)) {
        return jobResponse.data as T;
      }

      // Job failed
      if (isJobFailed(status)) {
        const error = jobResponse.error || "Job failed";
        throw new DevUtilsError(
          "JOB_FAILED",
          typeof error === "string" ? error : error.message || "Job failed",
          400,
        );
      }

      // Job still pending, continue polling
      pollAttempt++;
    } catch (error) {
      // If it's a 404, job might not exist yet (race condition)
      // Retry polling
      if (error && typeof error === "object" && (error as any).status === 404) {
        pollAttempt++;
        continue;
      }

      // Other errors should be thrown
      throw normalizeError(error);
    }
  }
}

/**
 * Resolve multiple jobs in parallel
 */
export async function resolveJobs<T = any>(
  responses: JobResponse[],
  pollFn: (jobId: string) => Promise<JobResponse>,
  config: Partial<JobResolverConfig> = {},
): Promise<T[]> {
  return Promise.all(
    responses.map((response) => resolveJob<T>(response, pollFn, config)),
  );
}

/**
 * Create a job resolver for a specific endpoint
 * Useful for reusing the same polling logic
 */
export function createJobResolver(
  pollFn: (jobId: string) => Promise<JobResponse>,
  config: Partial<JobResolverConfig> = {},
) {
  return {
    resolve: <T = any>(response: JobResponse) =>
      resolveJob<T>(response, pollFn, config),
    resolveMany: <T = any>(responses: JobResponse[]) =>
      resolveJobs<T>(responses, pollFn, config),
  };
}
